iml-gui/crate/src/components/command_modal.rs

```diff
@@ -284,3 +548,638 @@ fn close_button() -> Node<Msg> {
+    )
+}
+
+fn extract_children_from_job(job: &Job0) -> (u32, Vec<u32>) {
+    (
+        job.id,
+        job.steps.iter().filter_map(|s| extract_uri_id::<Step>(s)).collect(),
+    )
+}
+
+const fn extract_children_from_step(step: &Step) -> (u32, Vec<u32>) {
+    (step.id, Vec::new()) // steps have no descendants
+}
+
+fn extract_wait_fors_from_job(job: &Job0) -> (u32, Vec<u32>) {
+    // interdependencies between jobs
+    (
+        job.id,
+        job.wait_for.iter().filter_map(|s| extract_uri_id::<Job0>(s)).collect(),
+    )
+}
+
+fn extract_sorted_keys<T>(hm: &HashMap<u32, T>) -> Vec<u32> {
+    let mut ids = hm.keys().copied().collect::<Vec<_>>();
+    ids.sort();
+    ids
+}
+
+fn convert_to_sorted_vec<T: Clone>(hm: &HashMap<u32, T>) -> Vec<T> {
+    extract_sorted_keys(hm).into_iter().map(|k| hm[&k].clone()).collect()
+}
+
+fn select_by_keys<T: Clone>(hm: &HashMap<u32, T>, keys: &[u32]) -> Vec<T> {
+    keys.iter()
+        .filter(|k| hm.contains_key(k))
+        .map(|k| hm[k].clone())
+        .collect()
+}
+
+fn convert_to_rich_hashmap<T: Clone>(
+    ts: Vec<T>,
+    extract: impl Fn(&T) -> (u32, Vec<u32>),
+) -> HashMap<u32, Arc<Rich<u32, T>>> {
+    ts.into_iter()
+        .map(|t| {
+            let (id, deps) = extract(&t);
+            (id, Arc::new(Rich { id, deps, inner: t }))
+        })
+        .collect()
+}
+
+/// NOTE: the slices must be sorted
+pub fn is_subset<T: PartialEq>(part: &[T], all: &[T]) -> bool {
+    let mut idx = 0;
+    for it in part {
+        while (idx < all.len()) && (&all[idx] != it) {
+            idx += 1;
+        }
+        if idx == all.len() {
+            return false;
+        }
+    }
+    true
+}
+
+impl Model {
+    fn clear(&mut self) {
+        self.commands.clear();
+        self.commands_view.clear();
+        self.jobs.clear();
+        self.jobs_graph.clear();
+        self.steps.clear();
+        self.steps_view.clear();
+        self.select = Default::default();
+    }
+
+    fn assign_commands(&mut self, cmds: &[Command]) {
+        self.commands = convert_to_rich_hashmap(cmds.to_vec(), extract_children_from_cmd);
+        let tuple = self.check_consistency(&self.select);
+        self.refresh_view(tuple);
+    }
+
+    fn assign_jobs(&mut self, jobs: &[Job0]) {
+        self.jobs = convert_to_rich_hashmap(jobs.to_vec(), extract_children_from_job);
+        let tuple = self.check_consistency(&self.select);
+        self.refresh_view(tuple);
+    }
+
+    fn assign_steps(&mut self, steps: &[Step]) {
+        self.steps = convert_to_rich_hashmap(steps.to_vec(), extract_children_from_step);
+        let tuple = self.check_consistency(&self.select);
+        self.refresh_view(tuple);
+    }
+
+    /// We perform the consistency check of the current collections
+    /// `self.commands`, `self.jobs` and `self.steps`.
+    /// The selection, if it is non-empty, places additional constraints.
+    fn check_consistency(&self, select: &Select) -> (bool, bool, bool) {
```

Comment:
```suggestion
    fn check_consistency(&self) -> (bool, bool, bool) {
```

(and use `&self.select`)

Reply:
Done (there was a test, that made multiple `select` attempts for one model instance, now it is gone)